src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala

```diff
@@ -0,0 +1,138 @@
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.dto._
+import org.apache.spark.rdd.RDD
+import com.microsoft.partnercatalyst.fortis.spark.sinks.cassandra.udfs._
+import org.apache.spark.streaming.Time
+
+import scala.util.{Failure, Success, Try}
+
+object CassandraEventsSink{
+  private val KeyspaceName = "fortis"
+  private val TableEvent = "events"
+  private val TableEventTopics = "eventtopics"
+  private val TableEventPlaces = "eventplaces"
+  private val TableEventBatches = "eventbatches"
+  private val CassandraFormat = "org.apache.spark.sql.cassandra"
+  private val CassandraRetryIntervalMS = 2000
+  private val CassandraMaxRetryAttempts = 3
+
+  def apply(dstream: DStream[FortisEvent], sparkSession: SparkSession): Unit = {
+      val batchStream = dstream.foreachRDD{ (eventsRDD, time: Time) => {
+        if(!eventsRDD.isEmpty) {
+          val batchid = UUID.randomUUID().toString
+          val fortisEvents = writeFortisEvents(eventsRDD, batchid)
+          val aggregators = Seq(new PopularPlacesAggregator, new PopularTopicAggregator).par
+          val session = SparkSession.builder().config(eventsRDD.sparkContext.getConf)
+            .appName(eventsRDD.sparkContext.appName)
+            .getOrCreate()
+
+          if(session != null){
+            registerUDFs(session)
+            val eventBatchDF = fetchEventBatch(batchid, fortisEvents, session)
+            writeEventBatchToEventTagTables(eventBatchDF, session)
+            aggregators.map(aggregator => {
+              aggregateEventBatch(eventBatchDF, session, aggregator)
+            })
+          }
+        }
+      }}
+  }
+
+  private def writeFortisEvents(events: RDD[FortisEvent], batchId: String ): RDD[Event] = {
+    writeEventRddToEventsTable(events, batchId)
+  }
+
+  private def registerUDFs(session: SparkSession): Unit ={
+    session.sqlContext.udf.register("MeanAverage", FortisUdfFunctions.MeanAverage)
+    session.sqlContext.udf.register("SumMentions", FortisUdfFunctions.OptionalSummation)
+    session.sqlContext.udf.register("SentimentWeightedAvg", SentimentWeightedAvg)
+  }
+
+  /*Writes a Dstream of FortisEvents into the events table*/
+  private def writeEventRddToEventsTable(eventsRDD: RDD[FortisEvent], batchid: String): RDD[Event] = {
+    val fortisEventsDF = eventsRDD.map(CassandraEventSchema(_, batchid))
+    fortisEventsDF.saveToCassandra(KeyspaceName, TableEvent, writeConf = WriteConf(ifNotExists = true))
+
+    fortisEventsDF
+  }
+
+  private def fetchEventBatch(batchid: String, events: RDD[Event], session: SparkSession): Dataset[Event] = {
+    import session.sqlContext.implicits._
+
+    val addedEventsDF = session.sqlContext.read.format(CassandraFormat)
+      .options(Map("keyspace" -> KeyspaceName, "table" -> TableEventBatches))
+      .load()
+
+    addedEventsDF.createOrReplaceTempView(TableEventBatches)
+    val ds = session.sqlContext.sql(s"select eventid, pipelinekey " +
```

Comment:
NB: you could just use a multiline string here to save yourself all those pluses and quote characters:

```scala
val ds = session.sqlContext.sql(s"""select eventid, pipelinekey
                                  | from ${TableEventBatches}
                                  | where batchid = '${batchid}'""".stripMargin)
```